include/nlohmann/detail/iterators/internal_iterator.hpp

```diff
@@ -1,25 +1,372 @@
+                break;
+            case base_t::iterator_type::primitive:
+                ::new (&this->primitive_iterator()) primitive_iterator_t(o.primitive_iterator());
+                break;
+            case base_t::iterator_type::none:
+                break;
+            default:                 // LCOV_EXCL_LINE
+                JSON_ASSERT(false);  // NOLINT(cert-dcl03-c,hicpp-static-assert,misc-static-assert) LCOV_EXCL_LINE
+        }
+    }
+
+    void move_construct(internal_iterator_impl&& o) noexcept
+    {
+        m_it_type = o.m_it_type;
+        switch (m_it_type)
+        {
+            case base_t::iterator_type::object:
+                ::new (&this->object_iterator()) ObjectIterator(std::move(o.object_iterator()));
+                break;
+            case base_t::iterator_type::array:
+                ::new (&this->array_iterator()) ArrayIterator(std::move(o.array_iterator()));
+                break;
+            case base_t::iterator_type::primitive:
+                ::new (&this->primitive_iterator()) primitive_iterator_t(std::move(o.primitive_iterator()));
+                break;
+            case base_t::iterator_type::none:
+                break;
+            default:                 // LCOV_EXCL_LINE
+                JSON_ASSERT(false);  // NOLINT(cert-dcl03-c,hicpp-static-assert,misc-static-assert) LCOV_EXCL_LINE
+        }
+    }
+
+    typename base_t::iterator_type m_it_type;
+};
+
+/*!
+@brief Implementation of internal_iterator for trivially copyable types
 */
-template<typename BasicJsonType> struct internal_iterator
+template<class ObjectIterator, class ArrayIterator>
+class internal_iterator_impl<ObjectIterator, ArrayIterator, true> :
+    public internal_iterator_base<ObjectIterator, ArrayIterator>
 {
-    /// iterator for JSON objects
-    typename BasicJsonType::object_t::iterator object_iterator {};
-    /// iterator for JSON arrays
-    typename BasicJsonType::array_t::iterator array_iterator {};
-    /// generic iterator for all other types
-    primitive_iterator_t primitive_iterator {};
+  private:
+    using base_t = internal_iterator_base<ObjectIterator, ArrayIterator>;
+
+  public:
+    internal_iterator_impl() noexcept
+        : base_t(base_t::iterator_type::none)
+    {
+    }
+
+    explicit internal_iterator_impl(value_t v_type) noexcept
+        : base_t(base_t::iterator_type_from_value(v_type))
+    {
+    }
+
+    template<class ObjIt>
+    void set_object_iterator(ObjIt&& it) noexcept
+    {
+        this->object_iterator() = std::forward<ObjIt>(it);
+    }
+
+    template<class ArrIt>
+    void set_array_iterator(ArrIt&& it) noexcept
+    {
+        this->array_iterator() = std::forward<ArrIt>(it);
+    }
 };
+
+/*!
+@brief an iterator value
+*/
+template<class ObjectIterator, class ArrayIterator>
+using internal_iterator = internal_iterator_impl < ObjectIterator, ArrayIterator,
+#if defined(__clang__) || !defined(JSON_HEDLEY_GNUC_VERSION) || JSON_HEDLEY_GNUC_VERSION_CHECK(5, 0, 0)
```

Comment:
This condition isn't needed anymore, right? It's handled in your type trait implementation.